cw-orch-daemon/src/senders/sign.rs

```diff
@@ -0,0 +1,145 @@
+    },
+    CosmTxResponse, DaemonError, QuerySender, TxBuilder, TxSender,
+};
+use cosmrs::{
+    proto::cosmos::authz::v1beta1::MsgExec,
+    tendermint::chain::Id,
+    tx::{Body, Fee, Raw, SignDoc, SignerInfo},
+    AccountId, Any,
+};
+use prost::Message;
+
+pub struct SigningAccount {
+    pub account_number: u64,
+    pub sequence: u64,
+}
+
+pub trait CosmosSigner: QuerySender<Error = DaemonError> + Sync {
+    // --- General information about the signer --- //
+    /// The chain id of the connected chain
+    fn chain_id(&self) -> String;
+
+    /// The account id of the signer.
+    /// This is `_` prefixed because this conflict wih the TxSender trat
+    fn _account_id(&self) -> AccountId;
+
+    fn signing_account(
+        &self,
+    ) -> impl std::future::Future<Output = Result<SigningAccount, DaemonError>> + Send;
+
+    /// Signals wether this signer is using authz
+    /// If set to true, the signed messages will be wrapped inside authz messages
+    fn has_authz(&self) -> bool;
+
+    // --- Related to transaction signing --- //
+    /// Transaction signing
+    fn sign(&self, sign_doc: SignDoc) -> Result<Raw, DaemonError>;
+
+    fn signer_info(&self, sequence: u64) -> SignerInfo;
+
+    fn build_fee(&self, amount: impl Into<u128>, gas_limit: u64) -> Result<Fee, DaemonError>;
+
+    fn gas_price(&self) -> Result<f64, DaemonError>;
+
+    /// Computes the gas needed for submitting a transaction
+    fn calculate_gas(
+        &self,
+        tx_body: &Body,
+        sequence: u64,
+        account_number: u64,
+    ) -> impl std::future::Future<Output = Result<u64, DaemonError>> + Send {
+        async move {
+            let fee = self.build_fee(0u8, 0)?;
+
+            let auth_info = self.signer_info(sequence).auth_info(fee);
+
+            let sign_doc = SignDoc::new(
+                tx_body,
+                &auth_info,
+                &Id::try_from(self.chain_id())?,
+                account_number,
+            )?;
+
+            let tx_raw = self.sign(sign_doc)?;
+
+            Node::new_async(self.channel())
+                ._simulate_tx(tx_raw.to_bytes()?)
+                .await
+        }
+    }
+    /// Transaction broadcasting for Tendermint Transactions
+    fn broadcast_tx(
+        &self,
+        tx: Raw,
+    ) -> impl std::future::Future<
+        Output = Result<cosmrs::proto::cosmos::base::abci::v1beta1::TxResponse, DaemonError>,
+    > + Send {
+        async move {
+            let mut client = cosmos_modules::tx::service_client::ServiceClient::new(self.channel());
+            let commit = client
+                .broadcast_tx(cosmos_modules::tx::BroadcastTxRequest {
+                    tx_bytes: tx.to_bytes()?,
+                    mode: cosmos_modules::tx::BroadcastMode::Sync.into(),
+                })
+                .await?;
+
+            let commit = commit.into_inner().tx_response.unwrap();
+            Ok(commit)
+        }
+    }
+}
+
+impl<T: CosmosSigner + Sync> TxSender for T {
```

Comment:
It looks like having the trait bound `TxSender: Signer` makes sense!

Reply:
A TxSender can be anything that is used to interact with a Daemon. It could be a MultisigSender (no signing capabilities), it could be a QuerySender.. So Being a Signer is not a prerequisite for being a sender. However being a Signer makes you a sender

Reply:
MultisigSender would still need a key to do signing.

Reply:
Otherwise it's impossible to vote or do proposals